src/lib/car-index.js

```diff
@@ -147,3 +151,71 @@ export class StreamingCarIndex {
 }
 
 const mhToKey = (/** @type {Uint8Array} */ mh) => base58btc.encode(mh)
+
+export class BlocklyIndex {
+  /** R2 bucket where indexes live. */
+  #bucket
+  /** Cached index entries. */
+  #cache
+  /** Indexes that have been read. */
+  #indexes
+
+  /**
+   * @param {import('../bindings').R2Bucket} indexBucket
+   */
+  constructor (indexBucket) {
+    this.#bucket = indexBucket
+    /** @type {Map<RawCIDString, IndexEntry>} */
+    this.#cache = new Map()
+    this.#indexes = new Set()
+  }
+
+  /** @param {UnknownLink} cid */
+  async get (cid) {
+    // console.log(`index get: ${cid}`)
+    const key = cidToKey(cid)
+    let indexItem = this.#cache.get(key)
+    if (indexItem) {
+      // console.log(`index cache HIT: ${indexItem.origin}: ${cid} @ ${indexItem.offset}`)
+      if (cid.code !== raw.code) {
+        await this.#readIndex(cid)
+      }
+    } else {
+      await this.#readIndex(cid)
+      indexItem = this.#cache.get(key)
+      if (!indexItem) return // weird huh!?
+    }
+    return { cid, ...indexItem }
+  }
+
+  /**
+   * @param {import('multiformats').UnknownLink} cid
+   */
+  async #readIndex (cid) {
+    const key = cidToKey(cid)
+    if (this.#indexes.has(key)) return
+
+    // console.log(`reading block index: ${key}`)
+    const res = await this.#bucket.get(`${key}/${key}.idx`)
+    if (!res) return
+
+    const reader = MultiIndexReader.createReader({ reader: res.body.getReader() })
+    while (true) {
+      const { done, value } = await reader.read()
+      if (done) break
+
+      if (!('multihash' in value)) throw new Error('not MultihashIndexSorted')
+      const entry = /** @type {IndexEntry} */(value)
+      const rawCid = Link.create(raw.code, entry.multihash)
+      // console.log(`${cid}: ${value.origin}: ${rawCid} @ ${value.offset}`)
+      this.#cache.set(cidToKey(rawCid), entry)
+    }
+    this.#indexes.add(key)
+  }
+}
+
+/** @returns {RawCIDString} */
+const cidToKey = (/** @type {import('multiformats').UnknownLink} */ cid) => {
```

Comment:
nit: maybe call this `normalizeCid` or similar instead? Key makes it look like bucket keys